feat: add early stopping to Pregel

SemyonSinchenko opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. Ensure you have added or run the appropriate tests for your PR
  2. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]Your PR title ...'.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If there is design documentation, please add the link.
  3. If there is a discussion in the mailing list, please add the link.
-->
I added a `breakable` -> `break` to the `Pregel`. In the case of convergence before reaching `maxIter` it will give a significant performance benefit. In the case of not convergence before reaching `maxIter` it may introduce a small overhead, but because `Dataset.isEmpty` is not an action but is a quite cheap operation, the performance degradation is very slow. Another strong benefit to merge it is that allows users to set a very high value of `maxIter` if they are sure that their Pregel should converge, so it simplify development and usage.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
As described in #549 / #368

Close #549 
Close #368

SemyonSinchenko commented:
@SauronShepherd could you please take a look? It is exactly as we discussed in email. I also created #553 and I'm going to write a deep dive into how to use Pregel, so I will describe `earlyStopping` vs `maxIter` as part of my work on #553. Does it sounds good for you?
cc: @rjurney

SauronShepherd commented:
**Some Thoughts:**
1. It looks like you're executing the filter `msgDF.filter(Pregel.msg.isNotNull)` twice—once to check if it's empty (to stop the loop) and again to retrieve new messages. This transformation should be performed only once.
2. Why is the `checkpointDir` check inside the loop when it doesn't depend on the iteration? Also, why was this block committed under an unrelated topic (Spark Connect)?
3. `msgDF` is declared as a `var`, but it should be a `val` since it is never updated.
4. `setMaxIter` is the setter for `maxIter`, and `setCheckpointInterval` is the setter for `checkpointInterval`. For consistency, shouldn't `setEarlyStopping` be the setter for `earlyStopping` instead of the current `doEarlyStopping`?
5. Caching after checkpointing the DataFrame may be redundant.
6. Is `newVertexUpdateColDF.count()` really necessary?
7. Since the execution plan isn't that large, I think we should set the `checkpointInterval` to 5.

**Other Observations:**
- `sendMsgs` should also be a `val`.
- `CHECKPOINT_NAME_PREFIX` is not used.

**More Thoughts:**
`isEmpty` is actually an action and not a cheap operation, especially when **AQE** is enabled. I ran executions using your unit test without caching, unpersisting, or counting DataFrames, and the tests completed in 48 seconds without early stopping (but with `maxIter` reduced to 9 and `checkpointInterval` increased to 5, so only one checkpoint was performed). I also believe we could further improve the overall performance of this algorithm.

I think your "early stopping" feature is a great addition, particularly for cases where users don't know or can't estimate the required number of iterations, to prevent setting `maxIter` much higher than necessary. However, **it may introduce some overhead**, so when `maxIter` is low, it should be discouraged. We could mention this in the documentation, for example.

All that being said, I don’t believe any of my comments prevent this PR from being approved. We can merge it now and revisit it in the future.

SemyonSinchenko commented:
Thanks for the review!

> It looks like you're executing the filter msgDF.filter(Pregel.msg.isNotNull) twice—once to check if it's empty (to stop the loop) and again to retrieve new messages. This transformation should be performed only once.

Nice catch, thanks! Fixed!

> Why is the checkpointDir check inside the loop when it doesn't depend on the iteration? Also, why was this block committed under an unrelated topic (Spark Connect)?

It was added with a SparkConnect PR. It is not related anyhow to the current PR and it is marked as a change because I wrap all the `whil` into `breakable {}`. Fixed!

> msgDF is declared as a var, but it should be a val since it is never updated.

Fixed (actually it was before this PR)!

> setMaxIter is the setter for maxIter, and setCheckpointInterval is the setter for checkpointInterval. For consistency, shouldn't setEarlyStopping be the setter for earlyStopping instead of the current doEarlyStopping?

Fixed!

> Caching after checkpointing the DataFrame may be redundant.

Let's leave it for the further PRs? Otherwise it is hard to track performance improvements...

> Is newVertexUpdateColDF.count() really necessary?

I don't think so, but let's leave it for the next PR?

> We could mention this in the documentation, for example.

I added it to both scala docstring and python docstring:
```scala
  /**
   * Should Pregel stop earlier in case of no new messages to send?
   *
   * Early stopping allows to terminate Pregel before reaching maxIter by checking is there any
   * non-null message or not. While in some cases it may gain significant performance boost, it
   * other cases it can tend to performance degradation, because checking is messages DataFrame is
   * empty or not is an action and requires materialization of the Spark Plan with some additional
   * computations.
   *
   * In the case when user can assume a good value of maxIter it is recommended to leave this
   * value to the default "false". In the case when it is hard to estimate an amount of iterations
   * required for convergence, it is recommended to set this value to "false" to avoid iterating
   * over convergence until reaching maxIter. When this value is "true", maxIter can be set to a
   * bigger value without risks.
   *
   * @param value
   *   should Pregel checks for the termination condition on each step
   * @return
   */
```

I have also plans to write a comprehensive guide about Pregel.

> Since the execution plan isn't that large, I think we should set the checkpointInterval to 5.

It may fail in the case when there were some transformations in `edges` because we are joining with it on each iteration... Let's leave it for the next PRs?

SauronShepherd commented:
I'll have to add some minor changes in the Pregel class because of #554, but I'll keep the `cache` too.

SemyonSinchenko commented:
@SauronShepherd Let's merge this one to unblock your work on #554 and avoid merge conflicts?

SauronShepherd commented:
I don't see any other option but commenting on this issue.

SemyonSinchenko commented:
@rjurney Hi! Could you take a look and if everything is good, could you please merge it?

SemyonSinchenko commented:
@SauronShepherd I cannot get an answer from @rjurney, could you please review it? I need it to finalize the new proposed version of the "no-graphx" ShortestPaths.

rjurney commented:
@SemyonSinchenko sorry about that, I opened it to review last night. On it.

rjurney reviewed (approved):
lgtm

SemyonSinchenko commented on python/tests/test_graphframes.py:
```diff
@@ -207,6 +207,51 @@ def test_page_rank(spark):
+            * sqlfunctions.lit(1.0 - alpha)
+            + sqlfunctions.lit(alpha / numVertices),
+        )
+        .sendMsgToDst(pregel.src("rank") / pregel.src("outDegree"))
```
It is a naive implementation of PageRank (no voting for early stopping, no cases for the undirected graph, etc.)

One of the easiest Pregel-based algorithm, very easy to test.

rjurney replied:
Got it, I've been trying to program in pregel in my head for years but didn't see that :)
